sysdeps/managarm/include/mlibc/posix-pipe.hpp

```diff
@@ -278,12 +298,39 @@ auto exchangeMsgsSync(HelHandle descriptor, Args &&...args) {
 	);
 
 	auto element = globalQueue.dequeueSingle();
-	void *ptr = element.data();
+	__ensure(element);
+	void *ptr = element->data();
+
+	[&]<size_t... p>(std::index_sequence<p...>) {
+		(results.template get<p>().parse(ptr, *element), ...);
+	}(std::make_index_sequence<std::tuple_size_v<decltype(results)>>{});
+
+	return results;
+}
+
+template <typename... Args>
+auto exchangeMsgsSyncCancellable(HelHandle descriptor, HelHandle event, Args &&...args) {
+	auto results = helix_ng::createResultsTuple(args...);
+	auto actions = helix_ng::chainActionArrays(args...);
+
+	setCurrentRequestEvent(event);
+	HEL_CHECK(
+	    helSubmitAsync(descriptor, actions.data(), actions.size(), globalQueue.getQueue(), 0, 0)
+	);
+
+	auto element = globalQueue.dequeueSingle(false);
+	if (!element) {
+		element = globalQueue.dequeueSingle();
+		__ensure(element);
+	}
```

Comment:
This is the only call where we pass `false` to `dequeueSingle()` (AFAICT) and we immediately call into `dequeueSingle()` again. So what is the difference compared to just doing `dequeueSingle(true)` here?

Reply:
Note we use protocol errors to signify cancellation (e.g., in `read()` you check for `managarm::fs::Errors::INTERRUPTED`), so maybe the `ignoreCancel = false` case is not necessary (and we can always re-try if `helFutexWait` is interrupted)?